openshift/default-catalog-consistency/pkg/extract/extract.go

```diff
@@ -0,0 +1,180 @@
+package extract
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	imgcopy "github.com/containers/image/v5/copy"
+	"github.com/containers/image/v5/docker"
+	"github.com/containers/image/v5/manifest"
+	"github.com/containers/image/v5/oci/layout"
+	"github.com/containers/image/v5/pkg/compression"
+	"github.com/containers/image/v5/signature"
+	"github.com/containers/image/v5/types"
+	"github.com/containers/storage/pkg/archive"
+	"oras.land/oras-go/v2"
+	"oras.land/oras-go/v2/content/oci"
+)
+
+// Result represents the extracted OCI content and related context
+type Result struct {
+	Store  oras.ReadOnlyTarget
+	TmpDir string
+}
+
+// Cleanup function to remove the temporary directory
+func (r *Result) Cleanup() {
+	if r.TmpDir != "" {
+		if err := os.RemoveAll(r.TmpDir); err != nil {
+			fmt.Printf("failed to remove temp dir: %v\n", err)
+		}
+	}
+}
+
+// UnpackImage pulls the image, extracts it to disk, and opens it as an OCI store.
+func UnpackImage(ctx context.Context, imageRef, name string) (res *Result, err error) {
+	tmpDir, err := os.MkdirTemp("", fmt.Sprintf("oci-%s-", name))
+	if err != nil {
+		return nil, fmt.Errorf("create temp dir: %w", err)
+	}
+
+	res, err = func() (*Result, error) {
+		if err := pullImage(ctx, imageRef, tmpDir); err != nil {
+			return nil, fmt.Errorf("pull image: %w", err)
+		}
+
+		if err := getLayers(ctx, tmpDir); err != nil {
+			return nil, fmt.Errorf("extract filesystem: %w", err)
+		}
+
+		store, err := oci.New(tmpDir)
+		if err != nil {
+			return nil, fmt.Errorf("open OCI store: %w", err)
+		}
+
+		return &Result{
+			Store:  store,
+			TmpDir: tmpDir,
+		}, nil
+	}()
+
+	if err != nil {
+		os.RemoveAll(tmpDir)
+	}
+
+	return res, err
+}
+
+// pullImage pulls an image and writes it to OCI layout at the specified directory
+func pullImage(ctx context.Context, imageRef, tmpDir string) error {
+	srcRef, err := docker.ParseReference("//" + imageRef)
+	if err != nil {
+		return fmt.Errorf("parse image ref: %w", err)
+	}
+
+	destRef, err := layout.ParseReference(fmt.Sprintf("%s:%s", tmpDir, "v1"))
```

Comment:
Ah, I see. We are hardcoding to "v1" here. To me, this is pretty confusing. Do we need a tag at all for the layout reference?

Here's how we do this in the pull implementation shared between catalogd and operator-controller: https://github.com/openshift/operator-framework-operator-controller/blob/a2fd7089db389282bb370b8ae483f5d2764fd454/internal/shared/util/image/pull.go#L123-L126